src/components/TabList.tsx

```diff
@@ -0,0 +1,169 @@
+import { Div, DivProps, Flex, FlexProps } from 'honorable'
+import { AriaTabListProps } from '@react-types/tabs'
+import { Item } from '@react-stately/collections'
+import { useTab, useTabList, useTabPanel } from '@react-aria/tabs'
+import { TabListState } from '@react-stately/tabs'
+import { ItemProps, Node } from '@react-types/shared'
+import {
+  ComponentPropsWithRef,
+  HTMLAttributes,
+  RefObject,
+  useRef,
+} from 'react'
+
+import Tab from './Tab'
+
+type TabListStateProps = AriaTabListProps<object>;
+
+type Renderer = (
+  props: HTMLAttributes<HTMLElement>,
+  ref: RefObject<any>,
+  state: TabListState<object>
+) => JSX.Element;
+
+type MakeOptional<Type, Key extends keyof Type> = Omit<Type, Key> &
+  Partial<Pick<Type, Key>>;
+
+type TabListItemProps = ComponentPropsWithRef<typeof Tab> &
+  MakeOptional<ItemProps<void>, 'children'> & {
+    renderer?: Renderer;
+  };
+
+const TabListItem = Item as (props: TabListItemProps) => JSX.Element
+
+type TabListProps = {
+  state: TabListState<object>;
+  stateProps: TabListStateProps;
+  renderer?: Renderer;
+};
+function TabList({
+  state,
+  stateProps,
+  renderer,
+  ...props
+}: TabListProps & FlexProps) {
+  stateProps = {
+    ...{
+      keyboardActivation: 'manual',
+      orientation: 'horizontal',
+    },
+    ...stateProps,
+  }
+  const ref = useRef<HTMLDivElement>(null)
+  const { tabListProps } = useTabList(stateProps, state, ref)
+  const tabChildren = [...state.collection].map(item => (
+    <TabRenderer
+      key={item.key}
+      item={item}
+      state={state}
+      stateProps={stateProps}
+    />
+  ))
+  if (renderer) {
+    return renderer(
+      { ...props, ...tabListProps, ...{ children: tabChildren } },
+      ref,
+      state
+    )
+  }
+
+  return (
+    <Flex
+      {...tabListProps}
+      {...props}
+      flexDirection={stateProps.orientation === 'vertical' ? 'column' : 'row'}
+      alignItems={
+        stateProps.orientation === 'vertical' ? 'flex-start' : 'flex-end'
+      }
+      ref={ref}
+    >
+      {tabChildren}
+    </Flex>
+  )
+}
+
+type TabRendererProps = {
+  item: Node<unknown>;
+  state: TabListState<object>;
+  stateProps: TabListStateProps;
+};
+function TabRenderer({ item, state, stateProps }: TabRendererProps) {
+  const ref = useRef(null)
+  const { tabProps: props } = useTab({ key: item.key }, state, ref)
+
+  if (item.props.renderer) {
```

Comment:
This is equal to 
```
if (item.props.renderer && item.rendered) {}
```

Reply:
True, I haven't noticed that part. Then it would basically be:
```typescript
if (item.props.renderer) {
  return item.props.renderer(
        {
          ...{
            cursor: 'pointer',
            _focusVisible: { outline: '1px solid border-outline-focused' },
          },
          ...props,
        },
        ref,
        state
      )
}

if (item.rendered) {
      props.children = (
        <Tab
          active={state.selectedKey === item.key}
          vertical={stateProps.orientation === 'vertical'}
          width={stateProps.orientation === 'vertical' ? '100%' : 'auto'}
          {...item.props}
        >
          {item.rendered}
        </Tab>
      )
}
```

Reply:
Yes, the props.children update needs to happen before renderer() is called, and only if renderer exists.